astroML/utils/decorators.py

```diff
@@ -59,23 +61,25 @@ def new_f(*args, **kwargs):
             try:
                 D = pickle.load(open(filename, 'rb'))
                 cache_exists = True
-            except:
+            except Exception:
```

Comment:
a more specific Exception would be great, and below

Reply:
This is a tough one - depending on what the original author wanted to actually accomplish.  

I think it's supposed to check if a cached file exist and if not - create it. If the file exists the `D` dictionary (:roll_eyes:) gets populated by whatever pickle reads from it. I rewrote it in a try-except-else block because I believe that best reflects this, I assume intended, purpose.

At the same time it makes the caching system less dependable. If the cached file was corrupted, or if the function that was being cached returns some kind of complicated object that does imports, has external dependencies etc. the `pickle.load` function will fail. 

But pickle can fail in any number of ways, from `PickleError` to `KeyErrors`, `EOFError`, `ImportError` etc. - depending on what was pickled. In the original writeup, using a bare `except`, would catch any exception and thus cover both cases - if pickle failed and if the file didn't exist. Using `Exception just brings this particular intent closer to PEP8 standards. 

Would you prefer I returned to the original implementation or continue with the try-except-else?

Maybe there's an extra way, something I'm not aware of. I can, for example, wrap the load into a try-except for certain pickle-related fails but propagate others? Overall, however, this looks less readable to me than the original. 

-------------------------------------
As for "and bellow". The issue could be that we depend on the boolean equal comparison. So as long as we are in Python-builtin world that will work: 

```
>>> [1, 2, 3] == (1, 2, 3)
False
>>> [1, 2, 3] == 1
False
>>> [1, 2, 3] == "a"
False
>>> {"a":1, "b":2} == {"a":1, "b":2}
True
>>> {"a":1, "b":2} == {"a":1, "b":[2, 3]}
False
```
but 
```
>>> import numpy as np
>>> {"a":1, "b":2} == {"a":1, "b":np.array([2, 3])}
*** ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

There are many crazy scientists out there raising all kinds of comparison errors on custom objects. If those objects are put in the dictionaries technically any kind of error can propagate upwards. The code reads as if only builtin types or numpy types will be compared so I think `ValueError` is a very safe bet so that's what I put in. I removed comments and `TODO`'s.

Reply:
While I like your solution, I agree that the original was more broader and fell back to the behaviour of downloading/generating the data if there are any errors with the loading the pickle. So, I suppose keeping it as an Exception is fine. Marbe leave a comment that we do that as there are multiple ways this could fail.